state/application.go

```diff
@@ -2221,9 +2236,42 @@ func (a *Application) ChangeScale(scaleChange int) (int, error) {
 				{"unitcount", a.doc.UnitCount},
 				{"scale", a.doc.DesiredScale},
 			},
-			Update: bson.D{{"$set", bson.D{{"scale", newScale}}}},
+			Update: update,
 		}}
 
+		oldStorageConstraints, err := a.StorageConstraints()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		oldConstraints, err := a.Constraints()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		if scaleChange > 0 && len(attachStorage) > 0 {
+			for unitSeq := a.doc.UnitCount; unitSeq < newScale; unitSeq++ {
+				var virtualHostKey []byte
+				// We require a distinct virtual host key for each CAAS unit.
+				virtualHostKey, err = sshkeys.NewMarshalledED25519()
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+				// Specific unitName to avoid get unexpected unit sequence.
+				unitName := a.doc.Name + "/" + strconv.Itoa(unitSeq)
+				_, unitOps, err := a.addUnitOpsWithCons(applicationAddUnitOpsArgs{
+					unitName:       &unitName,
+					cons:           oldConstraints,
+					storageCons:    oldStorageConstraints,
+					attachStorage:  attachStorage,
+					VirtualHostKey: virtualHostKey,
+				})
```

Comment:
We shouldn't be doing all this here. It's just about attaching existing storage which has already been created. Storage constraints etc are not relevant (the storage already exists). Separate concerns like virtualHostKey are also not relevant (I'm pretty sure virtualHostKey is added via UpsertCAASUnit mentioned above.